pkg/controller/core/resource_controller.go

```diff
@@ -54,10 +54,60 @@ const (
+// of the reconcile function explicit.
 var (
-	Result        = reconcile.Result{}
-	ResultRequeue = reconcile.Result{Requeue: true}
+	// DeprecatedResultDone indicates that no further reconcilations need to be
+	// queued for the object under reconciliation. It is deprecated and should
+	// be replaced with a more explicitly named reconcile result.
+	DeprecatedResultDone = reconcile.Result{Requeue: false}
+
+	// DeprecatedResultRequeue indicates that further reconciliations need to be
+	// queued for the object under reconciliattion. It is deprecated and should
+	// be replaced with a more explicitly named reconcile result.
+	DeprecatedResultRequeue = reconcile.Result{Requeue: true}
+
+	// ResultDirty should be returned when the reconciler has altered the object
+	// under reconciliation. When returning this result the reconciler must also
+	// submit the updated object to the API server, thus triggering an implicit
+	// reconciliation of the changed object.
+	ResultDirty = reconcile.Result{Requeue: false}
+
+	// ResultDeleted should be returned when the object under reconciliation has
+	// been deleted, i.e. after the reconciler has removed any finalizers.
+	ResultDeleted = ResultDirty
+
+	// ResultClean should be returned when the reconciler has not altered the
+	// object under reconciliation, and is not waiting for a pending operation
+	// to complete.
+	ResultClean = reconcile.Result{RequeueAfter: RequeueOnSuccess}
+
+	// ResultWaiting should be returned when the reconciler has not altered the
+	// object under reconciliation, but is waiting for a pending operation to
+	// complete, for example waiting for the managed cloud resource the object
+	// represents to transition from 'Creating' to 'Available'.
+	ResultWaiting = reconcile.Result{RequeueAfter: RequeueOnWait}
+
+	// ResultNotFound should be returned when the object under reconciliation no
+	// longer exists when the reconcile function is invoked.
+	ResultNotFound = reconcile.Result{Requeue: false}
+
+	// ResultError should be returned if the reconciler knows it will return an
+	// error, thus triggering an implicit reconciliation.
+	ResultError = reconcile.Result{Requeue: false}
+
+	// ResultFailed should be returned if the reconciler had reason to set the
+	// object under reconciliation's status to 'failed'. This typically means
+	// an error was encountered and reflected in the object's status instead of
+	// being returned by the reconciler. In this scenario we want to requeue
+	// with exponential backoff. The eagle eyed reader may note that returning
+	// ResultFailed will duplicate the implicit reconcile triggered by setting
+	// the object's status to failed. We return ResultFailed rather than
+	// ResultDirty to account for the case in which the object was already, and
+	// remains, in state failed and thus does not trigger an implicit
+	// reconciliation.
+	ResultFailed = reconcile.Result{Requeue: true}
```

Comment:
I think there is a value for `reconcile.Result ` to represent reconciliation result, not in terms of the object state, but rather in terms of how to handle object "requeueing", granted that any result will be pre-empted by error value if there is one. 
For example:
```go
ReconcileDone = reconcile.Result{Requeue: false} // object does not need any further requeue
ReconcileSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
ReconcileWait    = reconcile.Result{RequeueAfter: requeueAfterOnWait}
ReconcileFailure = reconcile.Result{Requeue: true}
```
What do you think about this approach?

Reply:
I was considering this as the other angle too. My only concern is that it's not super obvious how "wait" is different from "success" or "done", however I think this will result in less awkward code branches as you mentioned, and we can clarify use cases using GoDoc.

Reply:
@ichekrygin PTAL. I was thinking about this while working on the Gitlab controller and landed on the same set of variables you suggested here, but named them in what I hope is a slightly more generic fashion while still being slightly intuitive.